File: crates/ferrite-core/src/types/connection.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum DatabaseDialect {
    PostgreSQL,
    SQLite,
}

impl std::fmt::Display for DatabaseDialect {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::PostgreSQL => write!(f, "postgresql"),
            Self::SQLite => write!(f, "sqlite"),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SslMode {
    Disable,
    Prefer,
    Require,
}

impl Default for SslMode {
    fn default() -> Self {
        Self::Prefer
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionConfig {
    pub id: Uuid,
    pub name: String,
    pub dialect: DatabaseDialect,
    pub host: Option<String>,
    pub port: Option<u16>,
    pub database_name: Option<String>,
    pub username: Option<String>,
    #[serde(skip_serializing)]
    pub password: Option<String>,
    pub ssl_mode: SslMode,
    pub color: Option<String>,
    pub sort_order: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionCreate {
    pub name: String,
    pub dialect: DatabaseDialect,
    pub host: Option<String>,
    pub port: Option<u16>,
    pub database_name: Option<String>,
    pub username: Option<String>,
    pub password: Option<String>,
    pub ssl_mode: Option<SslMode>,
    pub color: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionUpdate {
    pub name: Option<String>,
    pub host: Option<String>,
    pub port: Option<u16>,
    pub database_name: Option<String>,
    pub username: Option<String>,
    pub password: Option<String>,
    pub ssl_mode: Option<SslMode>,
    pub color: Option<String>,
    pub sort_order: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionTestResult {
    pub ok: bool,
    pub message: String,
    pub latency_ms: u64,
}

/// Resolved connection parameters — no URL encoding, passed directly to drivers.
#[derive(Debug, Clone)]
pub struct ConnectParams {
    pub dialect: DatabaseDialect,
    pub host: String,
    pub port: u16,
    pub database: String,
    pub username: String,
    pub password: Option<String>,
    pub ssl_mode: String,
}

File: app/src/renderer/types/connection.ts
export type DatabaseDialect = 'postgresql' | 'mysql' | 'sqlite'

export interface Connection {
  id: string
  name: string
  dialect: string
  host: string | null
  port: number | null
  database_name: string | null
  username: string | null
  ssl_mode: string
  color: string | null
  sort_order: number
  created_at: string
  updated_at: string
  connected: boolean
}

export interface ConnectionCreate {
  name: string
  dialect: DatabaseDialect
  host?: string
  port?: number
  database_name?: string
  username?: string
  password?: string
  ssl_mode?: string
  color?: string
}

export interface ConnectionUpdate {
  name?: string
  host?: string
  port?: number
  database_name?: string
  username?: string
  password?: string
  ssl_mode?: string
  color?: string
  sort_order?: number
}

export interface ConnectionTestResult {
  ok: boolean
  message: string
  latency_ms: number
}
